tests/Doctrine/Tests/DoctrineTestCase.php

```diff
@@ -5,10 +5,51 @@
 namespace Doctrine\Tests;
 
 use PHPUnit\Framework\TestCase;
+use PHPUnit\Runner\Version;
 
 /**
  * Base testcase class for all Doctrine testcases.
  */
 abstract class DoctrineTestCase extends TestCase
 {
+    public static function assertDoesNotMatchRegularExpression(
+        string $pattern,
+        string $string,
+        string $message = ''
+    ): void {
+        // Forward-compatibility wrapper for phpunit 9 : can be removed once phpunit 8 / php 7.2 support is dropped.
```

Comment:
I don't like traits much, but maybe this would be a good use case, that way we don't repeat that comment, and removing the methods means removing a file.

Reply:
Thinking about it more, maybe it'd be better to use class_alias and/or a conditional include so that the methods are only defined if running on the older phpunit version. Then there's no risk of conflict with any future phpunit version and the polyfills are isolated and easy to delete as above.

Conceptually with a trait it could be:

```
if (Version::series() === 8) {
  abstract class DoctrineTestCase extends TestCase {
    use Phpunit8ForwardCompatibilityPolyfill;
  }
} else {
  abstract class DoctrineTestCase extends TestCase {
  }
}
```

Although I'm not wild about swapping the `DoctrineTestCase` itself in case that needs methods added in future, would probably be better to do something at the `TestCase` level - maybe introduce an intermediate class a bit like this (not tested yet):

```php
// tests/Doctrine/Tests/PolyfillableTestCase.php
if (Version::series() === 8) {
  class_alias('Doctrine\Tests\PhpUnit8PolyfilledTestCase', 'Doctrine\Tests\PolyfillableTestCase');
} else {
  class_alias('PhpUnit\Framework\TestCase', 'Doctrine\Tests\PolyfillableTestCase');
}

// tests/Doctrine/Tests/DoctrineTestCase.php
abstract class DoctrineTestCase extends PolyfillableTestCase {
}
```

This has the advantage that you can easily delete the PHPUnit 8 polyfill when you drop support (presumably relatively soon, since it and PHP 7.2 are EOL?), but keep the mechanism so you can easily add e.g. a PHPUnit 9 polyfill when you upgrade to version 10. Yet is totally transparent (bar an extra class name in the hierarchy) for the latest supported version each time.

Reply:
Yes, that does look perfect, thanks 👍

Reply:
@acoulton i have simplified this into the `DoctrineTestCase` and using `__call` and `__callStatic` for #8613 for a few reasons:

- This is only about 1/3 of the code, and since we are "only in test code" here, we don't need to go over the top with the perfect abstraction. I prefer "fewer elements" when the solution is otherwise comparable.
- this is the same way we do it in DBAL, so code would be easier to copy/adapt/share
- its similarly easy to remove or extend this code when we change the PHPUnit requirements and since we will support several PHPUnit versions anyways, this kind of abstraction will always be present, having it in one place is better and `DoctrineTestCase` was introduced for this reason way back.